### ISSUE TITLE
Default configuration of Kibana (dashboards, indexes)

### ISSUE THREAD

**Issue by toszo** [epiphany-platform/epiphany#126]
Default dashboards can be populated by Filebeat. 
Look at: 
https://www.elastic.co/guide/en/beats/filebeat/current/load-kibana-dashboards.html

And remember that current installation of Kibana (OpenDistro) requires credentials to do requests.

**Comment by sbbroot** [comment]
Filebeat comes packaged with example Kibana dashboards, visualizations, and searches for visualizing Filebeat data in Kibana.

Filebeat configuration file is [here](https://github.com/epiphany-platform/epiphany/blob/develop/core/src/epicli/data/common/ansible/playbooks/roles/filebeat/templates/filebeat.yml.j2).

Default dashboards can be enabled under this [section](https://github.com/epiphany-platform/epiphany/blob/develop/core/src/epicli/data/common/ansible/playbooks/roles/filebeat/templates/filebeat.yml.j2#L178), with:
```
setup.dashboards.enabled: true
```

Before that an index called 'filebeat-*' needs to be created. This is the default name, can be changed by using:
```
setup.dashboards.index: value
```

Things to consider:
- Kibana needs to be running, if not filebeat will crash with an error. To avoid this, these arguments could be used:
```
setup.dashboards.retry.enabled: true  # if Kibana is not reachable filebeat will try to reconnect.
setup.dashboards.retry.interval  # 1 second by default.
setup.dashboards.retry.maximum  # unlimited by default.
```
- Filebeat has multiple choices when it comes to choosing a path to dashboards (we might consider user input).

**Comment by przemyslavic** [comment]
@sbbroot 
@to-bar (as you can be helpful in this topic 😏)

:heavy_check_mark: Works fine for new clusters. I tested different values for `specification.kibana.dashboards.enabled` (`auto`/`true`/`false`) and `specification.kibana.dashboards.index` parameters. The only non-working case causing Filebeat to crash (as expected) is when dashboards are set to true and Kibana is disabled.

❌ However, there is a problem with the upgrades. While the upgrade from v1.0 and v1.1 works fine, the installation fails when you try to upgrade from v0.8 and v0.9. It fails on task `filebeat : Wait for filebeat service to be running`.
```
2021-07-20T17:18:43.8245353Z[38;21m17:18:43 INFO cli.engine.ansible.AnsibleCommand - TASK [filebeat : Wait for filebeat service to be running] **********************
2021-07-20T17:18:46.7891836Z[31;21m17:18:46 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (10 retries left).
2021-07-20T17:18:46.8139279Z[31;21m17:18:46 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (10 retries left).
2021-07-20T17:18:50.7187214Z[31;21m17:18:50 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (9 retries left).
2021-07-20T17:18:50.8767342Z[31;21m17:18:50 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (9 retries left).
2021-07-20T17:18:54.6131961Z[31;21m17:18:54 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (8 retries left).
2021-07-20T17:18:54.7750691Z[31;21m17:18:54 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (8 retries left).
2021-07-20T17:18:58.5240376Z[31;21m17:18:58 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (7 retries left).
2021-07-20T17:18:58.7211511Z[31;21m17:18:58 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (7 retries left).
2021-07-20T17:19:02.4888671Z[31;21m17:19:02 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (6 retries left).
2021-07-20T17:19:03.0015324Z[31;21m17:19:02 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (6 retries left).
2021-07-20T17:19:06.4398021Z[31;21m17:19:06 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (5 retries left).
2021-07-20T17:19:07.1255883Z[31;21m17:19:07 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (5 retries left).
2021-07-20T17:19:10.4894723Z[31;21m17:19:10 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (4 retries left).
2021-07-20T17:19:11.1406151Z[31;21m17:19:11 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (4 retries left).
2021-07-20T17:19:14.4484800Z[31;21m17:19:14 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (3 retries left).
2021-07-20T17:19:15.1968518Z[31;21m17:19:15 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (3 retries left).
2021-07-20T17:19:18.4253777Z[31;21m17:19:18 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (2 retries left).
2021-07-20T17:19:19.1146771Z[31;21m17:19:19 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (2 retries left).
2021-07-20T17:19:22.3657182Z[31;21m17:19:22 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (1 retries left).
2021-07-20T17:19:23.4017200Z[31;21m17:19:23 ERROR cli.engine.ansible.AnsibleCommand - FAILED - RETRYING: Wait for filebeat service to be running (1 retries left).
2021-07-20T17:19:26.3129698Z[31;21m17:19:26 ERROR cli.engine.ansible.AnsibleCommand - fatal: [ec2-1-1-1-1.eu-west-1.compute.amazonaws.com]: FAILED! => {"attempts": 10, "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
2021-07-20T17:19:27.5754856Z[31;21m17:19:27 ERROR cli.engine.ansible.AnsibleCommand - fatal: [ec2-2-2-2-2.eu-west-1.compute.amazonaws.com]: FAILED! => {"attempts": 10, "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
```

```
filebeat.service: Main process exited, code=exited, status=1/FAILURE
filebeat.service: Failed with result 'exit-code'.
filebeat.service: Service hold-off time over, scheduling restart.
filebeat.service: Scheduled restart job, restart counter is at 5.
Stopped Filebeat sends log files to Logstash or directly to Elasticsearch..
filebeat.service: Start request repeated too quickly.
filebeat.service: Failed with result 'exit-code'.
Failed to start Filebeat sends log files to Logstash or directly to Elasticsearch..
```

It looks like it can't connect to Kibana, especially since it's trying to query localhost and not Kibana vm.
Perhaps some settings were lost during the upgrade.

```
root@ec2-2-2-2-2:~# journalctl -f -t filebeat
-- Logs begin at Tue 2021-07-20 16:27:56 UTC. --
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.016Z        WARN        [tls]        tlscommon/tls_config.go:83        SSL/TLS verifications disabled.
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.016Z        INFO        [publisher]        pipeline/module.go:113        Beat name: ec2-2-2-2-2.eu-west-1.compute.amazonaws.com
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.017Z        INFO        kibana/client.go:119        Kibana url: http://localhost:5601
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.017Z        INFO        [monitoring]        log/log.go:118        Starting metrics logging every 30s
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.021Z        INFO        [monitoring]        log/log.go:153        Total non-zero metrics        {"monitoring": {"metrics": {"beat":{"cpu":{"system":{"ticks":0,"time":{"ms":3}},"total":{"ticks":40,"time":{"ms":44},"value":40},"user":{"ticks":40,"time":{"ms":41}}},"handles":{"limit":{"hard":4096,"soft":1024},"open":8},"info":{"ephemeral_id":"c8415016-9e6a-4d0e-8026-cce1a61306d5","uptime":{"ms":32}},"memstats":{"gc_next":9007792,"memory_alloc":6103504,"memory_total":11063472,"rss":44916736},"runtime":{"goroutines":10}},"filebeat":{"harvester":{"open_files":0,"running":0}},"libbeat":{"config":{"module":{"running":0}},"output":{"type":"elasticsearch"},"pipeline":{"clients":0,"events":{"active":0}}},"registrar":{"states":{"current":0}},"system":{"cpu":{"cores":2},"load":{"1":1.12,"15":0.73,"5":0.69,"norm":{"1":0.56,"15":0.365,"5":0.345}}}}}}
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.022Z        INFO        [monitoring]        log/log.go:154        Uptime: 35.065323ms
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.022Z        INFO        [monitoring]        log/log.go:131        Stopping metrics logging.
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.022Z        INFO        instance/beat.go:447        filebeat stopped.
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: 2021-07-20T17:18:44.023Z        ERROR        instance/beat.go:951        Exiting: error connecting to Kibana: fail to get the Kibana version: HTTP GET request to http://localhost:5601/api/status fails: fail to execute the HTTP GET request: Get "http://localhost:5601/api/status": dial tcp [::1]:5601: connect: connection refused. Response: .
Jul 20 17:18:44 ec2-2-2-2-2.eu-west-1.compute.amazonaws.com filebeat[15447]: Exiting: error connecting to Kibana: fail to get the Kibana version: HTTP GET request to http://localhost:5601/api/status fails: fail to execute the HTTP GET request: Get "http://localhost:5601/api/status": dial tcp [::1]:5601: connect: connection refused. Response: .
```

Kibana config before upgrade:

```
cat /etc/filebeat/filebeat.yml.bak_7.8.1
#============================== Kibana =====================================

# Starting with Beats version 6.0.0, the dashboards are loaded via the Kibana API.
# This requires a Kibana endpoint configuration.
# setup.kibana:

#   # Kibana Host
#   # Scheme and port can be left out and will be set to the default (http and 5601)
#   # In case you specify and additional path, the scheme is required: http://localhost:5601/path
#   # IPv6 addresses should always be defined as: https://[2001:db8::1]:5601
# #   host: "10.1.3.202:5601"
# #   username: logstash
#   password: logstash

  # Kibana Space ID
  # ID of the Kibana Space into which the dashboards should be loaded. By default,
  # the Default Space will be used.
  #space.id:
```

After upgrade:
```
cat /etc/filebeat/filebeat.yml

#============================== Kibana =====================================

# Starting with Beats version 6.0.0, the dashboards are loaded via the Kibana API.
# This requires a Kibana endpoint configuration.
setup.kibana:

  # Kibana Host
  # Scheme and port can be left out and will be set to the default (http and 5601)
  # In case you specify and additional path, the scheme is required: http://localhost:5601/path
  # IPv6 addresses should always be defined as: https://[2001:db8::1]:5601
  #host: "localhost:5601"

  # Kibana Space ID
  # ID of the Kibana Space into which the dashboards should be loaded. By default,
  # the Default Space will be used.
  #space.id:
```

It is possible that this is a general problem with the Filebeat upgrade. But it has only now been discovered.
Moving back to TODO to investigate the issue.

EDIT after investigating with @to-bar 
In v1.0 there were some changes introduced to the Kibana configuration in the Filebeat template (https://github.com/epiphany-platform/epiphany/blob/v1.0/core/src/epicli/data/common/ansible/playbooks/roles/filebeat/templates/filebeat.yml.j2#L226). In v0.8 and v0.9 the entire Kibana section is commented out (https://github.com/epiphany-platform/epiphany/blob/v0.8/core/src/epicli/data/common/ansible/playbooks/roles/filebeat/templates/filebeat.yml.j2#L217), which means there is no config provided for Kibana in the Filebeat config file.
During Filebeat upgrade to develop, the existing Filebeat config is loaded to read some data (https://github.com/epiphany-platform/epiphany/blob/develop/core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/filebeat.yml#L102), but since the Kibana section in 0.8 and 0.9 is commented out, the configuration for Kibana is actually empty, and due to the dashboard loading enabled in this PR, Filebeat tries to communicate with Kibana using the default address `localhost:5601`, which results in the error described above.

**Comment by przemyslavic** [comment]
It looks like `existing_setup_dashboards` variable is never used (https://github.com/epiphany-platform/epiphany/pull/2467/files#diff-72a446ee42566d641402eb05aedbe6e97ba54a6a93532ce284e6785ac648ddc7R102).
Probably it should be used in some places in the template instead of the `existing_setup_kibana` variable.

**Comment by przemyslavic** [comment]
There are still issues with preserving some settings in the configuration file when upgrading filebeat. Please take a look at @to-bar's review https://github.com/epiphany-platform/epiphany/pull/2467#pullrequestreview-731745377.

**Comment by przemyslavic** [comment]
Fixed in #2467.
Tested `epicli apply` and `epicli upgrade` from v0.8. User settings are preserved when upgrading.
Also verified aarch64 offline installation.
